pnpm/test/install/overrides.ts

```diff
@@ -0,0 +1,74 @@
+import path from 'path'
+import { preparePackages } from '@pnpm/prepare'
+import writeYamlFile from 'write-yaml-file'
+import { execPnpm } from '../utils'
+
+test('overrides with local file and link specs', async () => {
+  const projects = preparePackages([
+    {
+      location: '.',
+      package: {
+        name: 'root',
+      },
+    },
+    {
+      location: 'packages/main',
+      package: {
+        name: 'main',
+        dependencies: {
+          'relative-file-pkg': '*',
+          'absolute-file-pkg': '*',
+          'relative-link-pkg': '*',
+          'absolute-link-pkg': '*',
+        },
+      },
+    },
+    {
+      location: 'overrides/pkg',
+      package: {
+        name: 'pkg',
+        version: '0.0.0',
+      },
+    },
+  ])
+
+  // overrides must be written after prepare because it depends on cwd
+  projects.root.writePackageJson({
+    name: 'root',
+    pnpm: {
+      overrides: {
+        'relative-file-pkg': 'file:./overrides/pkg',
+        'absolute-file-pkg': `file:${path.resolve('overrides/pkg')}`,
+        'relative-link-pkg': 'link:./overrides/pkg',
+        'absolute-link-pkg': `link:${path.resolve('overrides/pkg')}`,
+      },
+    },
+  })
+
+  await writeYamlFile('pnpm-workspace.yaml', { packages: ['packages/*'] })
+
+  await execPnpm(['install'])
+
+  const lockfile = projects.root.readLockfile()
+
+  expect(lockfile.importers['packages/main']).toStrictEqual({
+    dependencies: {
+      'relative-file-pkg': {
+        specifier: 'file:../../overrides/pkg',
+        version: 'pkg@file:overrides/pkg',
+      },
+      'absolute-file-pkg': {
+        specifier: `file:${path.resolve('overrides/pkg')}`,
+        version: 'pkg@file:overrides/pkg',
+      },
+      'relative-link-pkg': {
+        specifier: 'link:../../overrides/pkg',
+        version: 'link:../../overrides/pkg',
+      },
+      'absolute-link-pkg': {
+        specifier: `link:${path.resolve('overrides/pkg')}`,
+        version: 'link:../../overrides/pkg',
+      },
+    },
+  } as typeof lockfile['importers'][string])
+})
```

Comment:
@zkochan So I have written 4 test cases (the "dependency" cases require modifying `registry-mock` and shall be done later). And I see that the `version` fields in the lockfile are probably not correct. What do you think should be the correct values? Should the versions of absolute paths also absolute? And is it intended that `link:` doesn't have `pkg@` like `file:`?

Reply:
> And is it intended that link: doesn't have pkg@ like file:?

Yes, but `file:` also shouldn't have `pkg@` if `pkg` is the real name in the `package.json` of `pkg`.

> Should the versions of absolute paths also absolute?

That should be OK to keep it absolute. absolute overrides doesn't makes sense in any case. I am not sure how users use them. Probably they don't commit those changes.